docs/contributing/999-rfcs/2024-02-15-resource-pinning.md

```diff
@@ -0,0 +1,239 @@
+---
+title: Resource Pinning
+description: Protection of resources from accidental deletion/moves
+---
+
+# Resource Pinning
+
+- **Author(s):**: @MarkMcCulloh
+- **Submission Date**: 2024-02-15
+- **Stage**: Proposal
+- **Stage Date**: 2024-02-15
+
+Introduce a mechanism to protect resources from accidental deletion/moves in the logical application tree.
+
+## Requirements
+
+### Functional
+- User-ability in sources to signify intent for a resource to not be deleted or moved (e.g. "pinning")
+- Provide default for a resource's "pinned" state in it's implementation
+- Git-trackable way to store changes to pinned resources between compilations, per target platform (e.g. "pinfile")
+- User-ability to manually and declaratively override a pinned resource's path in the resource tree
+- User-ability to manually and declaratively override a pinned resource's attributes used by the platform
+- Compilation failure when a pinned resource is deleted without explicit removal from the pinfile
+- Compilation failure when a pinned resource is moved without explicit mapping of a previous path to a new path
+- Compilation failure when a resource is no longer pinned in source but is still present in the pinfile
+
+### Non-Functional
+- The pinfile is human-readable and editable
+- The pinfile contains the minimal information necessary to achieve the functional requirements
+- The pinfile is stable and predictable between compilations on different machines
+- CLI provides clear error messages and guidance for resolving deletion/moves of pinned resources
+
+### Cases
+
+For the following cases, assume a user has expressed their intent to protect a resource from deletion/moves somehow.
+The following cases should be handled by this RFC:
+
+- A pinned resource's id was changed in source (using `as`)
+- A pinned resource was moved to a different path in the resource tree (e.g. placed inside a different class)
+- An update to wing itself changed a pinned resource's path
+- An update to a dependency changed the id's within a resource it exports, and the user has pinned an instance of that resource
+
+## User Journey
+
+Sally Stateful creates a wing file with a bucket
+
+```wing
+// main.w
+bring cloud;
+
+new cloud.Bucket() as "CoolBucket";
+```
+
+They run `wing compile` and then `wing test`. Nice, looks like everything works.
+
+Running `wing compile -t tf-aws` also succeeds with the following output and a new file called main.w.pin.json appears:
+
+```shell
+# (output)
+[+pin] /root/CoolBucket
+```
+
+```json
+// main.w.pin.json
+{
+  "version": "1",
+  "pinned": {
+    "tf-aws": {
+      "/root/CoolBucket": {
+        "type": "aws_s3_bucket",
+        "uniqueId": "CoolBucket",
```

Comment:
```suggestion
        "terraformId": "CoolBucket",
```

I think this should be explicit

Reply:
Removed this property for now, not sure if it's actually useful